### DISCUSSION TITLE
IOS Crash MGLLineStyleLayer when use predicate

### DISCUSSION THREAD

**Discussion by jojonay** [maplibre/maplibre-native#1623] [Q&A]
My app crashes after using the layer.predicate method for example
` mylayer.predicate = NSPredicate(format: "%K != %@", "mykey", "")`

The app can continue to work. and the display of lines and map movements. Still able to display normally But when I zoomed in, somewhere my app crashed. But when zooming in to areas where the lines are less dense, Maybe it can still work.

example code
```
let url = URL(string: "https://example.com/mapbox_traffic.json")!
let source = MGLVectorTileSource(identifier: "traffic", configurationURL: url)
mapView.style?.addSource(source)

let layerTraffic = MGLLineStyleLayer(identifier: "traffic-forword", source: source)
layerTraffic.sourceLayerIdentifier = "traffic"

layerTraffic.predicate = NSPredicate(format: "%K != %@", "mykey", "") 
mapView.style?.addLayer(layerTraffic)

```

**Comment by alanchenboy** [comment]
use expression-related things is a risk now. They will generate warnings if you use NSpredicate(format: or NSExpression(format: and even crash in some situations, I think we need a pan to refactor the whole expression wrap on the iOS side, just like Mapbox doing now. The only concern is that Mapbox implements the expression via Swift, If we want to migrate to Swift on the iOS side and then build an objective-c wrap is it reasonable?

**Comment by alanchenboy** [comment]
There is a workaround here `NSPredicate:(predicateWithMGLJSONObject:["!=", ["get", "mykey"], "" ])` I didn't test this expression but I think it will discard the format call to prevent it from the crash.
